include/boost/windows_sspi/detail/sspi_impl.hpp

```diff
@@ -17,11 +17,169 @@
+    InBuffers[0].cbBuffer = static_cast<ULONG>(m_input_data.size());
+    InBuffers[0].BufferType = SECBUFFER_TOKEN;
+
+    InBuffers[1].pvBuffer = nullptr;
+    InBuffers[1].cbBuffer = 0;
+    InBuffers[1].BufferType = SECBUFFER_EMPTY;
+
+    InBuffer.cBuffers = 2;
+    InBuffer.pBuffers = InBuffers;
+    InBuffer.ulVersion = SECBUFFER_VERSION;
+
+    OutBuffers[0].pvBuffer = nullptr;
+    OutBuffers[0].BufferType = SECBUFFER_TOKEN;
+    OutBuffers[0].cbBuffer = 0;
+
+    OutBuffer.cBuffers = 1;
+    OutBuffer.pBuffers = OutBuffers;
+    OutBuffer.ulVersion = SECBUFFER_VERSION;
+
+    m_last_error = detail::sspi_functions::InitializeSecurityContext(m_credentials,
+                                                                     m_context,
+                                                                     nullptr,
+                                                                     m_flags_in,
+                                                                     0,
+                                                                     SECURITY_NATIVE_DREP,
+                                                                     &InBuffer,
+                                                                     0,
+                                                                     nullptr,
+                                                                     &OutBuffer,
+                                                                     &m_flags_out,
+                                                                     nullptr);
+
+    if (InBuffers[1].BufferType == SECBUFFER_EXTRA) {
+      // Some data needs to be reused for the next call, move that to the front for reuse
+      // TODO: Test that this works.
+      std::move(m_input_data.end() - InBuffers[1].cbBuffer, m_input_data.end(), m_input_data.begin());
+      m_input_data.resize(InBuffers[1].cbBuffer);
+    } else {
+      m_input_data.clear();
+    }
+
+    if (OutBuffers[0].cbBuffer != 0 && OutBuffers[0].pvBuffer != nullptr) {
+      // TODO: Avoid this copy
+      m_output_data = std::vector<char>{reinterpret_cast<const char*>(OutBuffers[0].pvBuffer)
+        , reinterpret_cast<const char*>(OutBuffers[0].pvBuffer) + OutBuffers[0].cbBuffer};
+      detail::sspi_functions::FreeContextBuffer(OutBuffers[0].pvBuffer);
+      return state::data_available;
+    }
+
+    switch (m_last_error) {
+      case SEC_E_INCOMPLETE_MESSAGE:
+        return state::data_needed;
+
+      case SEC_E_OK:
+        BOOST_ASSERT_MSG(InBuffers[1].BufferType != SECBUFFER_EXTRA, "Handle extra data from handshake");
+        return state::done;
+
+      case SEC_I_INCOMPLETE_CREDENTIALS:
+        BOOST_ASSERT_MSG(false, "client authentication not implemented");
+
+      default:
+        return state::error;
+    }
+  }
+
+  // TODO: Consider making this more flexible by not requering a
+  // vector of chars, but any view of a range of bytes
```

Comment:
this would be a boost::span pre c++20, or a std::span >= c++20

Reply:
or to be compatible with asio & beast, a `net::const_buffer`

Reply:
My thoughts as well. Probably a `net::const_buffer` would make most sense, but as I tried to indicate in my comment, I'd prefer to look into that after I have improved my testing. In my experience, it just makes changing things like this a lot easier.